gradio/routes.py

```diff
@@ -462,6 +478,59 @@ def get_server_url_from_ws_url(ws_url: str):
 set_documentation_group("routes")
 
 
+class Obj:
+    """
+    Using a class to convert dictionaries into objects. Used by the `Request` class.
+    Credit: https://www.geeksforgeeks.org/convert-nested-python-dictionary-to-object/
+    """
+
+    def __init__(self, dict1):
+        self.__dict__.update(dict1)
+
+    def __str__(self) -> str:
+        return str(self.__dict__)
+
+    def __repr__(self) -> str:
+        return str(self.__dict__)
+
+
+@document()
+class Request:
+    """
+    A Gradio request object that can be used to access the request headers, cookies,
+    query parameters and other information about the request from within the prediction
+    function. The class is a thin wrapper around the fastapi.Request class.
+
+    Example:
+        import gradio as gr
+        def echo(name, request: gr.Request):
+            print("Request headers dictionary:", request.headers)
+            print("IP address:", request.client.host)
+            return name
+        io = gr.Interface(echo, "textbox", "textbox").launch()
+    """
+
+    def __init__(self, request: fastapi.Request | None = None, **kwargs):
+        """
+        Can be instantiated with either a fastapi.Request or by manually passing in
+        attributes (needed for queueing).
+        """
+        self.request = request
+        self.kwargs = kwargs
+
+    def dict_to_obj(self, d):
+        if isinstance(d, dict):
+            return json.loads(json.dumps(d), object_hook=Obj)
+        else:
+            return d
+
+    def __getattr__(self, name):
+        if self.request:
+            return self.dict_to_obj(getattr(self.request, name))
```

Comment:
For consistency the request and websocket cases should both raise an AttributeError if the attribute doesn't exist or return None. Right now only the request case raises an AttributeError

Reply:
Nice agreed